libp2p-xtra/src/dialer.rs

```diff
@@ -0,0 +1,114 @@
+use crate::Connect;
+use crate::ConnectionStats;
+use crate::Endpoint;
+use crate::GetConnectionStats;
+use anyhow::anyhow;
+use async_trait::async_trait;
+use libp2p_core::Multiaddr;
+use std::time::Duration;
+use tokio_tasks::Tasks;
+use xtra::Address;
+use xtra_productivity::xtra_productivity;
+use xtras::SendInterval;
+
+/// How often we check whether we are still connected
+pub const CHECK_INTERVAL: Duration = Duration::from_secs(5);
+
+/// xtra actor that takes care of dialing (connecting) to an Endpoint.
+///
+/// Periodically polls Endpoint to check whether connection is still active.
+/// Should be used in conjunction with supervisor maintaining resilient connection.
+pub struct Actor {
+    tasks: Tasks,
+    endpoint: Address<Endpoint>,
+    connect_address: Multiaddr,
+    /// Contains the reason we are stopping.
+    stop_reason: Option<Error>,
+}
+
+impl Actor {
+    pub fn new(endpoint: Address<Endpoint>, connect_address: Multiaddr) -> Self {
+        Self {
+            tasks: Tasks::default(),
+            endpoint,
+            connect_address,
+            stop_reason: None,
+        }
+    }
+
+    async fn check_connection_active_in_endpoint(&self) -> Result<(), Error> {
```

Comment:
I am having a hard time following the code in here.
Let me try to summarize what I understand to validate that is is correct:

1. We use a supervisor actor to restart the `dialer::Actor` in case the actor emits a `dialer::Error`.
2. Inside the `dialer::Actor` we monitor the connection every `5` seconds

So, if handling `CheckConnection` in `check_connection_active_in_endpoint` fails we set the stop reason to `ConnectionDropped` and then stop the actor, which results in the actor being restarted by the supervisor. 

I think it would help me to inline the code of `check_connection_active_in_endpoint` in the handler so there is a bit less indirection.

Reply:
Yes, exactly!

Unfortunately, inlining messes up error handling. I've added a comment, hopefully it will be a bit easier to understand with it.